AlphaWallet/Tokens/Types/TokensDataStore.swift

```diff
@@ -577,21 +586,18 @@ class MultipleChainsTokensDataStore: NSObject, TokensDataStore {
         return false
     }
 
-    private func updateNonFungibleBalance(balance: NonFungibleBalance, token: TokenObject) -> Bool {
+    private func updateNonFungibleBalance(balance: NonFungibleBalance, token: TokenObject, realm: Realm) -> Bool {
         //NOTE: add new balances
         let balancesToAdd = balance.rawValue
             .filter { b in !token.balance.contains(where: { v in v.balance == b }) }
             .map { TokenBalance(balance: $0) }
 
         //NOTE: remove old balances if something has changed
-        let balancesToDelete = token.balance
-            .filter { !balance.rawValue.contains($0.balance) }
-            .compactMap { token.balance.index(of: $0) }
+        let balancesToDelete = Array(token.balance
+            .filter { !balance.rawValue.contains($0.balance) })
 
         if !balancesToAdd.isEmpty || !balancesToDelete.isEmpty {
-            for index in balancesToDelete {
-                token.balance.remove(at: index)
```

Comment:
The old version doesn't seem to work correctly because when it deletes the size (and hence the indices) are modified.

But I don't understand how both the old and new version of deletion works though. Why do we delete the `TokenBalance` instance from `token.balance` array if it matches?

Reply:
we can suppose that function `updateNonFungibleBalance` calls with valid balances values, so to avoid replacing values that not changed
we look for balances that we are going to delete, (that don't contain in `balance:` parameter), that might be old balances. To remove values we loopp through list, to resolve object index in list, as `List<>` has method `remove(at:` that accepts objects index. Then just map throught indecies to delete object. Out of this unchanged balances stays same, new balances will be added to list, old balances removed from list. There only one issue, that `remove(at:` method doesn't remove object from Realm, so it stays in realm.

Reply:
Oh I see!